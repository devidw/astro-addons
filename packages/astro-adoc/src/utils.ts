/**
 * Recursively get all .adoc files in a given directory
 */
import { promises as fs } from 'fs'
import { join } from 'path'

export async function getFiles(dir: string): Promise<string[]> {
    const dirents = await fs.readdir(dir, { withFileTypes: true })
    const files = await Promise.all(dirents.map((dirent) => {
        const res = join(dir, dirent.name)
        return dirent.isDirectory() ? getFiles(res) : res
    }))
    return Array.prototype.concat(...files)
}

export async function getAdocPaths(dir: string, paramName: string = "slug"): Promise<{ params: { [key: string]: string } }[]> {
    // maybe add trailing slash if not present
    if (!dir.endsWith('/')) {
        dir = dir + '/'
    }
    const files = await getFiles(dir)
    const adocFiles = files.filter((file) => file.endsWith('.adoc'))
    return adocFiles.map((file) => {
        let path = file.replace('.adoc', '')

        if (path.endsWith('/index')) {
            path = path.replace('/index', '')
        }

        path = path.replace(dir, '')

        return { params: { [paramName]: path } }
    })
}