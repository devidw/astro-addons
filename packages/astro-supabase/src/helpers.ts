import type { SupabaseClient, AuthChangeEvent, Session } from "@supabase/supabase-js"
import type { AstroGlobal } from "astro"
import Cookies from 'js-cookie'
import type { CookieAttributes } from 'js-cookie'

/**
 * supabase.auth.api.setAuthCookie()
 */
export function setAuthCookie(event: AuthChangeEvent, session: Session | null) {
  const cookieOptions: CookieAttributes = {
    sameSite: 'strict',
  }

  if (event === 'SIGNED_IN') {
    if (!session) throw new Error('Session is null')

    Cookies.set('sb-access-token', session.access_token, cookieOptions)

    if (session.refresh_token) {
      Cookies.set('sb-refresh-token', session.refresh_token)
    }
  }

  if (event === 'SIGNED_OUT') {
    ['sb-access-token', 'sb-refresh-token'].forEach((cookie) => {
      Cookies.remove(cookie)
    })
  }
}

/**
 * supabaseClient.auth.api.getUserByCookie()
 */
export async function getUserByCookie(Astro: AstroGlobal, supabaseClient: SupabaseClient) {
  // supabaseClient.auth.api.setAuthCookie
  // supabaseClient.auth.api.getUserByCookie
  const accessToken = Astro.cookies.get("sb-access-token")
  const refreshToken = Astro.cookies.get("sb-refresh-token")

  if (!accessToken.value) {
    return null
  }

  const { user, error: getUserError } = await supabaseClient.auth.api.getUser(
    accessToken.value
  )

  // if (getUserError) {
  //   if (!refreshToken.value) {
  //     return null
  //   }

  //   const { error } = await supabaseClient.auth.api.refreshAccessToken(
  //     refreshToken.value
  //   )

  //   if (error) {
  //     return null
  //   }
  // }

  return user
}