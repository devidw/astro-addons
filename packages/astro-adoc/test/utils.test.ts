import { expect, test } from 'vitest'
import { getAdocFiles, getAdocPaths, getAdoc } from '../src/utils'
// import { Asciidoctor } from 'asciidoctor'

test('getAdocFiles', async () => {
    const files = await getAdocFiles('test/fixtures')
    expect(files).toMatchSnapshot()
})

test('getAdocPaths', async () => {
    const paths = await getAdocPaths('test/fixtures')
    expect(paths).toMatchSnapshot()
})

test('getAdocPaths with custom parameter name', async () => {
    const paths = await getAdocPaths('test/fixtures', 'custom')
    expect(Object.keys(paths[0].params)).toEqual(['custom'])
})

test('get existing asciidoctor document', () => {
    const adoc = getAdoc('test/fixtures/document.adoc')
    // expect(adoc).toBeInstanceOf(Asciidoctor.Document)
    expect(adoc.getTitle()).toEqual('Document Title')
})

test('get non-existing asciidoctor document', () => {
    expect(() => getAdoc('test/fixtures/non-existing.adoc')).toThrow()
})

test('get index asciidoctor document', () => {
    const direct = getAdoc('test/fixtures/documents/index.adoc')
    const indirect = getAdoc('test/fixtures/documents')
    expect(direct.getTitle()).toEqual(indirect.getTitle())
})