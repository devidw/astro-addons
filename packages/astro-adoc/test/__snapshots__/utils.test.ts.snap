// Vitest Snapshot v1

exports[`getAdocFiles 1`] = `
[
  "test/fixtures/document.adoc",
  "test/fixtures/documents/document.adoc",
  "test/fixtures/documents/index.adoc",
]
`;

exports[`getAdocPaths 1`] = `
[
  {
    "params": {
      "slug": "document",
    },
  },
  {
    "params": {
      "slug": "documents/document",
    },
  },
  {
    "params": {
      "slug": "documents",
    },
  },
]
`;
